### PR TITLE
hardening: add NoNewPrivileges

### PR DESCRIPTION
**Pull request by miekg**
Add this to harden the unit being executed. Move the tmpfs slightly up,
so it fits better with the rest of the Fs options.

Signed-off-by: Miek Gieben <miek@miek.nl>

### CODE REVIEW

File: internal/provider/pod.go [hunk 1]
```diff
@@ -170,6 +170,9 @@ func (p *p) CreatePod(ctx context.Context, pod *corev1.Pod) error {
 		uf = uf.Overwrite("Service", "ProtectHome", "tmpfs")
 		uf = uf.Overwrite("Service", "PrivateMounts", "true")
 		uf = uf.Overwrite("Service", "ReadOnlyPaths", "/")
+		uf = uf.Overwrite("Service", "NoNewPrivileges", "true")
```
Comment (pires) [inline internal/provider/pod.go:173]: Why? I'm unsure whether this will break privileged Pods.

**Comment by miekg** [comment]
[ Quoting <notifications@github.com> in "Re: [virtual-kubelet/systemk] harde..." ]
>@pires commented on this pull request.
>
>━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
>
>In internal/provider/pod.go:
>
>> @@ -170,6 +170,9 @@ func (p *p) CreatePod(ctx context.Context, pod *corev1.Pod) error {
>                uf = uf.Overwrite("Service", "ProtectHome", "tmpfs")
>                uf = uf.Overwrite("Service", "PrivateMounts", "true")
>                uf = uf.Overwrite("Service", "ReadOnlyPaths", "/")
>+               uf = uf.Overwrite("Service", "NoNewPrivileges", "true")
>
>Why? I'm unsure whether this will break privileged Pods.

I would like to contrain the units as much as possible. Didn't consider privileged pods
though.

**Comment by pires** [comment]
I'm in favor of secure by default but we can't forget the Pod API.

**Comment by miekg** [comment]
[ Quoting <notifications@github.com> in "Re: [virtual-kubelet/systemk] harde..." ]
>I'm in favor of secure by default but we can't forget the Pod API.

Yeah, fair point. *Reading some docs*
(https://kubernetes.io/docs/concepts/policy/pod-security-policy/#privileged)

Seems privileged means access to /dev not gaining new privileges. Note that systemd allows
shielding of /dev/ (PrivateDevices) which currently is unused.

**Comment by pires** [comment]
This is more what I was referring to, sorry https://kubernetes.io/docs/concepts/policy/pod-security-policy/#privilege-escalation

**Comment by miekg** [comment]
[ Quoting <notifications@github.com> in "Re: [virtual-kubelet/systemk] harde..." ]
>This is more what I was referring to, sorry https://kubernetes.io/docs/concepts
>/policy/pod-security-policy/#privilege-escalation

ah, yes that different. This harks back to the discussion on how much of the API we can
possibly support and if we should detect unsupported options.

On that page the `AllowedCapabilies` is also something we could support.

I think we should draw a line on what it means to run under systemk wrt to these kind of
things.

**Comment by pires** [comment]
Well, we can also backtrack and implement this change when privilege escalation is explicitly denied in pod spec.

**Comment by miekg** [comment]
this needs a proper doc on how much access you get by default while running under systemk - there are many ways we can restrict this from the systemd side, but maybe not all of them make sense. Closing for now

**Pull request closed without merging**